src/sage/graphs/homomorphisms/count_homomorphisms.py

```diff
@@ -0,0 +1,236 @@
+
+        if colourful and (graph_clr is None or target_clr is None):
+            raise ValueError("Both graph_clr and target_clr must be provided when colourful is True")
+
+        self.graph._scream_if_not_simple()
+        self.target_graph._scream_if_not_simple()
+
+        self.tree_decomp = graph.treewidth(certificate=True)
+        self.nice_tree_decomp = make_nice_tree_decomposition(graph, self.tree_decomp)
+        self.root = sorted(self.nice_tree_decomp)[0]
+
+        # Make it into directed graph for better access
+        # to children and parent, if needed
+        #
+        # Each node in a labelled nice tree decomposition
+        # has the following form:
+        #
+        # (node_index, bag_vertices) node_type
+        #
+        # Example: (5, {0, 4}) intro
+        self.dir_labelled_TD = label_nice_tree_decomposition(self.nice_tree_decomp, self.root, directed=True)
+
+        # `node_changes_dict` is responsible for recording introduced and
+        # forgotten vertices in a nice tree decomposition
+        self.node_changes_dict = node_changes(self.dir_labelled_TD)
+
+        # `DP_table` is a vector/list of dictionaries
+        # Each element (dict) corresponds to the (induced) hom's of a tree node.
+        # For each pair inside a dict, the key is the hom, and the value is the number of hom's:
+        #
+        # An example of K2 to K3, the values are arbitrary for demo purpose:
+        #
+        # [{((4, 0), (5, 1)): 10,
+        #   ((4, 0), (5, 2)): 20,
+        #   ((4, 1), (5, 0)): 30,
+        #   ((4, 1), (5, 2)): 40,
+        #   ((4, 2), (5, 0)): 50,
+        #   ((4, 2), (5, 1)): 60}, {}, ...]
+        self.DP_table = [{} for _ in range(len(self.dir_labelled_TD))]
+
+    def count_homomorphisms(self):
+        r"""
+        Return the number of homomorphisms from the graph `G` to the graph `H`.
+
+        A homomorphism from a graph `G` to a graph `H` is a function
+        `\varphi : V(G) \mapsto V(H)`, such that for any edge `uv \in E(G)` the
+        pair `\varphi(u) \varphi(v)` is an edge of `H`.
+
+        For more information, see the :wikipedia:`Graph_homomorphism`.
+
+        ALGORITHM:
+
+        This is an implementation based on the proof of Prop. 1.6 in [CDM2017]_.
+
+        OUTPUT:
+
+        - an integer, the number of homomorphisms from `graph` to `target_graph`
+
+        EXAMPLES::
+
+            sage: from sage.graphs.homomorphisms.count_homomorphisms import GraphHomomorphismCounter
+            sage: square = graphs.CycleGraph(4)
+            sage: bip = graphs.CompleteBipartiteGraph(2, 4)
+            sage: counter = GraphHomomorphismCounter(square, bip)
+            sage: counter.count_homomorphisms()
+            128
+        """
+        # Whether it's BFS or DFS, every node below join node(s) would be
+        # computed already, so we can go bottom-up safely.
+        for node in reversed(self.dir_labelled_TD.vertices()):
```

Comment:
Can you clarify the order in which you want to visit the vertices. Here, you take the reverse order of the list returned by `vertices()` but `vertices()` does not ensure any kind of ordering.

Reply:
I just read the implementation code for `vertices()`, since `self.dir_labelled_TD` is a DAG, I assume in this case, the vertices returned would/should be from the source to the sink(s)?

Reply:
No, we cannot make any assumption on this ordering. It depends on internal data structures and not on the orientation of the graph. If you need a specific ordering, you must use an appropriate graph traversal method (BFS, DFS, etc.).